In [1]:
# import libraries

import numpy as np
import random
import math 


In [2]:
# method 1: numpy.cos(x)

# the derivative functin of sin(x) is cos(x), thus we use np.cos(x)
x = 1.0
exact = np.cos(x)
print (f"The result computed by np.cos(x) is: {exact}")


The result computed by np.cos(x) is: 0.5403023058681398


In [34]:
# method 2: 3-point endpoint formula:

def Three_PointEndpoint(x,h):
    derivative = ((-3*(np.sin(x)))+(4*(np.sin(x+h)))-(np.sin(x+(2*h))))/((2*h))
    return derivative

h=0.1
T_end_ary = []

# vary h to find the value of h that leads to the most accurate results
for i in range(100):
    T_end = Three_PointEndpoint(x,h)
    T_end_ary.append(T_end)
    
    print (i,f". Result from three end point is: {T_end}.")
    print (f"   When h is ",h)
    h = h/10



0 . Result from three end point is: 0.5418869992741304.
   When h is  0.1
1 . Result from three end point is: 0.5403201049504258.
   When h is  0.01
2 . Result from three end point is: 0.540302485758426.
   When h is  0.001
3 . Result from three end point is: 0.5403023076705482.
   When h is  0.0001
4 . Result from three end point is: 0.5403023059014078.
   When h is  1e-05
5 . Result from three end point is: 0.5403023059513677.
   When h is  1.0000000000000002e-06
6 . Result from three end point is: 0.540302307894258.
   When h is  1.0000000000000002e-07
7 . Result from three end point is: 0.5403023140004847.
   When h is  1.0000000000000002e-08
8 . Result from three end point is: 0.5403025249428594.
   When h is  1.0000000000000003e-09
9 . Result from three end point is: 0.5403033576101278.
   When h is  1.0000000000000003e-10
10 . Result from three end point is: 0.5403122393943248.
   When h is  1.0000000000000003e-11
11 . Result from three end point is: 0.5405120795387574.
   When 

In [40]:
# Determine the absolute and relative errors for each of your calculations
# we assume that using np.cos(x) as the exact result.

#abs  : current - exact
T_abs = []
#rel  : abs/exact
T_rel = []

h = 0.1
print("For three-point endpoint formula, the errors are shown as following when h is different: \n")

for i in range(len(T_end_ary)):
    a = abs(T_end_ary[i] - exact)
    r = a/exact
    T_abs.append(a)
    T_rel.append(r)
    print("\n","When h is " , h)
    print("Absolute error: ", a)
    print("Relative error: ", r)
    h=h/10


For three-point endpoint formula, the errors are shown as following when h is different: 


 When h is  0.1
Absolute error:  0.0015846934059906559
Relative error:  0.0029329754635128258

 When h is  0.01
Absolute error:  1.779908228605187e-05
Relative error:  3.294282125532094e-05

 When h is  0.001
Absolute error:  1.7989028622267256e-07
Relative error:  3.329437691990428e-07

 When h is  0.0001
Absolute error:  1.8024084447176847e-09
Relative error:  3.335925879164322e-09

 When h is  1e-05
Absolute error:  3.326805497749774e-11
Relative error:  6.157303904902596e-11

 When h is  1.0000000000000002e-06
Absolute error:  8.322798006332732e-11
Relative error:  1.540396536520409e-10

 When h is  1.0000000000000002e-07
Absolute error:  2.0261182731573513e-09
Relative error:  3.7499715458401605e-09

 When h is  1.0000000000000002e-08
Absolute error:  8.132344908595712e-09
Relative error:  1.5051471778431393e-08

 When h is  1.0000000000000003e-09
Absolute error:  2.1907471958737545e-07
Rel

In [43]:
smallest_err_T = min(T_abs)
exponent_T = -(T_abs.index(smallest_err_T))
th = 0.1 * math.pow(10,exponent_T)

smallest_err_T2 = min(T_rel)
exponent_T2 = -(T_rel.index(smallest_err_T2))
th2 = 0.1 * math.pow(10,exponent_T2)

print( "The smallest abslute error is ", smallest_err_T)
print( "When h is ", th)

print( "The smallest relative error is ", smallest_err_T2)
print( "When h is ", th2)

print("Thus, the most accurate result for this method is ", T_end_ary[-exponent_T])

The smallest abslute error is  3.326805497749774e-11
When h is  1e-05
The smallest relative error is  6.157303904902596e-11
When h is  1e-05
Thus, the most accurate result for this method is  0.5403023059014078


# Report for Three-point Endpoint Method :

According to the three-point endpoint formula from the class, I made a method called "Three_PointEndpoint". This method is able to estimate the derivative of $sin(x)$ given $x$ and $h$. And then I use a for-loop to vary $h$ so that we can find the value of h that leads to the most accurate results. The first 10 results of estimation are:

0 . Result from three end point is: 0.5418869992741304.<br>
   When h is  0.1<br>
1 . Result from three end point is: 0.5403201049504258.<br>
   When h is  0.01<br>
2 . Result from three end point is: 0.540302485758426.<br>
   When h is  0.001<br>
3 . Result from three end point is: 0.5403023076705482.<br>
   When h is  0.0001<br>
4 . Result from three end point is: 0.5403023059014078.<br>
   When h is  1e-05<br>
5 . Result from three end point is: 0.5403023059513677.<br>
   When h is  1.0000000000000002e-06<br>
6 . Result from three end point is: 0.540302307894258.<br>
   When h is  1.0000000000000002e-07<br>
7 . Result from three end point is: 0.5403023140004847.<br>
   When h is  1.0000000000000002e-08<br>
8 . Result from three end point is: 0.5403025249428594.<br>
   When h is  1.0000000000000003e-09<br>
9 . Result from three end point is: 0.5403033576101278.<br>
   When h is  1.0000000000000003e-10<br>

The $h$ is ranging from $0.1$ to $1.0*10^{-100}$. Here, I only show the first 10 results since they are closest to the exact result by observing the result list above. The corresponding absolute and relative errors for each calculations are:<br>

 When h is  0.1<br>
Absolute error:  0.0015846934059906559  |    Relative error:  0.0029329754635128258<br>

 When h is  0.01<br>
Absolute error:  1.779908228605187e-05  |  Relative error:  3.294282125532094e-05<br>

 When h is  0.001<br>
Absolute error:  1.7989028622267256e-07  |  Relative error:  3.329437691990428e-07<br>

 When h is  0.0001<br>
Absolute error:  1.8024084447176847e-09  |  Relative error:  3.335925879164322e-09<br>

 When h is  1e-05<br>
Absolute error:  3.326805497749774e-11  |  Relative error:  6.157303904902596e-11<br>

 When h is  1.0000000000000002e-06<br>
Absolute error:  8.322798006332732e-11  |  Relative error:  1.540396536520409e-10<br>

 When h is  1.0000000000000002e-07<br>
Absolute error:  2.0261182731573513e-09  |  Relative error:  3.7499715458401605e-09<br>

 When h is  1.0000000000000002e-08<br>
Absolute error:  8.132344908595712e-09  |  Relative error:  1.5051471778431393e-08<br>

 When h is  1.0000000000000003e-09<br>
Absolute error:  2.1907471958737545e-07  |  Relative error:  4.054669343588558e-07<br>

 When h is  1.0000000000000003e-10<br>
Absolute error:  1.0517419880562429e-06  |  Relative error:  1.9465806024394787e-06<br>

By using python's min() method, we are able to reach the smallest absolute and relative error from the error lists that I made above. The results are:<br>
The smallest abslute error is  3.326805497749774e-11<br>
When h is  1e-05<br>
The smallest relative error is  6.157303904902596e-11<br>
When h is  1e-05<br>

Thus, we can conclude that for Three-point Endpoiint method, when the $h$ is $1*10^{-5}$, we have the samllest error to the exact value, which is the most accurate result. And when $h$ is $1*10^{-5}$, the result of $sin^{'}(1.0)$ is 0.5403023059014078. 


In [47]:
# method 3: 5-point midpoint formula:

def Five_PointMidpoint(x,h):
    derivative = ((np.sin(x-(2*h)))-(8*(np.sin(x-h)))+(8*(np.sin(x+h)))-(np.sin(x+(2*h))))/((12*h))
    return derivative

h=0.1
F_mid_ary = []

#vary h to find the value of h that leads to the most accurate results
for i in range(100):
    F_mid = Five_PointMidpoint(x,h)
    F_mid_ary.append(F_mid)
    
    print (i,f".Result from five mid point is: {F_mid}.")
    print (f"   When h is ",h)
    h = h/10
    

0 .Result from five mid point is: 0.5403005070032606.
   When h is  0.1
1 .Result from five mid point is: 0.5403023056880424.
   When h is  0.01
2 .Result from five mid point is: 0.5403023058679901.
   When h is  0.001
3 .Result from five mid point is: 0.5403023058681936.
   When h is  0.0001
4 .Result from five mid point is: 0.5403023058634752.
   When h is  1e-05
5 .Result from five mid point is: 0.5403023058866048.
   When h is  1.0000000000000002e-06
6 .Result from five mid point is: 0.5403023059513677.
   When h is  1.0000000000000002e-07
7 .Result from five mid point is: 0.5403023065989979.
   When h is  1.0000000000000002e-08
8 .Result from five mid point is: 0.5403022936463958.
   When h is  1.0000000000000003e-09
9 .Result from five mid point is: 0.5403026174614447.
   When h is  1.0000000000000003e-10
10 .Result from five mid point is: 0.5402992867923708.
   When h is  1.0000000000000003e-11
11 .Result from five mid point is: 0.5402807830752938.
   When h is  1.00000000000000

In [32]:
# Determine the absolute and relative errors for each of your calculations
# we assume that using np.cos(x) as the exact result.

#abs  : current - exact
F_abs = []
#rel  : abs/exact
F_rel = []

h = 0.1
print("For five-point midpoint formula, the errors are shown as following when h is different: \n")

for i in range(len(F_mid_ary)):
    a = abs(F_mid_ary[i] - exact)
    r = a/exact
    F_abs.append(a)
    F_rel.append(r)
    print("\n","When h is " , h)
    print("Absolute error: ", a)
    print("Relative error: ", r)
    h=h/10
    

For five-point midpoint formula, the errors are shown as following when h is different: 


 When h is  0.1
Absolute error:  1.7988648791655848e-06
Relative error:  3.329367392343863e-06

 When h is  0.01
Absolute error:  1.800973814525264e-10
Relative error:  3.3332706430551304e-10

 When h is  0.001
Absolute error:  1.496580637194711e-13
Relative error:  2.769894966096906e-13

 When h is  0.0001
Absolute error:  5.384581669432009e-14
Relative error:  9.96586838692136e-14

 When h is  1e-05
Absolute error:  4.664602037962595e-12
Relative error:  8.633318768587648e-12

 When h is  1.0000000000000002e-06
Absolute error:  1.846500730096068e-11
Relative error:  3.4175325739711065e-11

 When h is  1.0000000000000002e-07
Absolute error:  8.322798006332732e-11
Relative error:  1.540396536520409e-10

 When h is  1.0000000000000002e-08
Absolute error:  7.308581517762036e-10
Relative error:  1.352683754702629e-09

 When h is  1.0000000000000003e-09
Absolute error:  1.2221743950213693e-08
Relativ

In [45]:
smallest_err_F = min(F_abs)
exponent_F = -(F_abs.index(smallest_err_F))
fh = 0.1 * math.pow(10,exponent_F)

smallest_err_F2 = min(F_rel)
exponent_F2 = -(F_rel.index(smallest_err_F2))
fh2 = 0.1 * math.pow(10,exponent_F2)

print( "The smallest abslute error is ", smallest_err_F)
print( "When h is ", fh)
print( "The smallest relative error is ", smallest_err_F2)
print( "When h is ", fh2)
print("Thus, the most accurate result for this method is ", F_mid_ary[-exponent_F])

The smallest abslute error is  5.384581669432009e-14
When h is  0.0001
The smallest relative error is  9.96586838692136e-14
When h is  0.0001
Thus, the most accurate result for this method is  0.5403023058681936


# Report for Five-point Midpoint Method :

According to the five-point midpoint formula from the class, I made a method called "Five_PointMidpoint". This method is able to estimate the derivative of $sin(x)$ given $x$ and $h$. And then I use a for-loop to vary $h$ so that we can find the value of h that leads to the most accurate results. The first 10 results of estimation are:<br>

0 .Result from five mid point is: 0.5403005070032606.<br>
   When h is  0.1<br>
1 .Result from five mid point is: 0.5403023056880424.<br>
   When h is  0.01<br>
2 .Result from five mid point is: 0.5403023058679901.<br>
   When h is  0.001<br>
3 .Result from five mid point is: 0.5403023058681936.<br>
   When h is  0.0001<br>
4 .Result from five mid point is: 0.5403023058634752.<br>
   When h is  1e-05<br>
5 .Result from five mid point is: 0.5403023058866048.<br>
   When h is  1.0000000000000002e-06<br>
6 .Result from five mid point is: 0.5403023059513677.<br>
   When h is  1.0000000000000002e-07<br>
7 .Result from five mid point is: 0.5403023065989979.<br>
   When h is  1.0000000000000002e-08<br>
8 .Result from five mid point is: 0.5403022936463958.<br>
   When h is  1.0000000000000003e-09<br>
9 .Result from five mid point is: 0.5403026174614447.<br>
   When h is  1.0000000000000003e-10<br>

The $h$ is ranging from $0.1$ to $1.0*10^{-100}$. Here, I only show the first 10 results since they are closest to the exact result by observing the result list above. The corresponding absolute and relative errors for each calculations are:<br>

When h is  0.1<br>
Absolute error:  1.7988648791655848e-06  |  Relative error:  3.329367392343863e-06<br>

 When h is  0.01<br>
Absolute error:  1.800973814525264e-10  |  Relative error:  3.3332706430551304e-10<br>

 When h is  0.001<br>
Absolute error:  1.496580637194711e-13  |  Relative error:  2.769894966096906e-13<br>

 When h is  0.0001<br>
Absolute error:  5.384581669432009e-14  |  Relative error:  9.96586838692136e-14<br>

 When h is  1e-05<br>
Absolute error:  4.664602037962595e-12  |  Relative error:  8.633318768587648e-12<br>

 When h is  1.0000000000000002e-06<br>
Absolute error:  1.846500730096068e-11  |  Relative error:  3.4175325739711065e-11<br>

 When h is  1.0000000000000002e-07<br>
Absolute error:  8.322798006332732e-11  |  Relative error:  1.540396536520409e-10<br>

 When h is  1.0000000000000002e-08<br>
Absolute error:  7.308581517762036e-10  |  Relative error:  1.352683754702629e-09<br>

 When h is  1.0000000000000003e-09<br>
Absolute error:  1.2221743950213693e-08  |  Relative error:  2.2620195800527264e-08<br>

 When h is  1.0000000000000003e-10<br>
Absolute error:  3.1159330493579773e-07  |  Relative error:  5.767017862993199e-07<br>


By using python's min() method, we are able to reach the smallest absolute and relative error from the error lists that I made above. The results are:<br>
The smallest abslute error is  5.384581669432009e-14<br>
When h is  0.0001<br>
The smallest relative error is  9.96586838692136e-14<br>
When h is  0.0001<br>

Thus, we can conclude that for Five-point Midpoint Method, when the $h$ is $0.0001$, we have the samllest error to the exact value, which is the most accurate result. And when $h$ is $0.0001$, the result of $sin^{'}(1.0)$ is 0.5403023058681936. 